### DISCUSSION TITLE
Sponsored issue: SSG pages should need to be refreshed one time after on-demand revalidation to see the update

### DISCUSSION THREAD

**Discussion by pingpongdoctor** [vercel/next.js#56298] [Help]
When employing SSG pages, the necessity to on-demand revalidate poses a requirement to observe updates following data modifications. Upon executing on-demand revalidation via revalidateTag or revalidatePath, a dual-step process unfolds, mandating two refresh intervals for updated data manifestation.

The initial refresh is dedicated to initiating the background execution of the revalidation process. Subsequently, the second refresh aligns with updating cached data, presenting pages featuring the latest information. This bifurcated refresh sequence impedes the immediate display of amendments, adversely impacting user experience.

I hereby seek assistance in optimizing the on-demand revalidation procedure to streamline the process, enabling the exhibition of updated data with a singular page refresh. This enhancement is pivotal for ensuring a seamless and responsive user experience, particularly for SSG pages requiring prompt updates upon data augmentation.

## Priority Support

- @pingpongdoctor is using  [Mintycode](https://mintycode.io) to fund this issue.
- If you would like to accept ![amount](https://imageupload.io/ib/zwkd2EaAY6YpPsg_1693209112.png) bounty for solving this issue join [Mintycode](https://mintycode.io/profile?action=CREATOR_SUPPORT_MODAL&requestId=b0178887-eb06-4b3a-ab27-fc4ec81e5db5&owner=vercel&name=next.js).
- Thank you in advance for helping.

[![mintycode](https://imageupload.io/ib/LvxMwrPxam1JW9e_1693210482.png)](https://mintycode.io)

**Comment by balazsorban44** [comment]
I converted this to a discussion. Please use our issue templates in the future.

**Comment by jwalcher** [comment]
The need for double refresh indeed is annoying. Note however that this affects "only" fetching of external data, while the pages themselves are recompiled at the first request following revalidatePath/revalidateTag. Also, the res.revalidate method that is provided in the pages directory and still works on routes in app directory does the full revalidation at once, including external data fetching. So perhaps the behavior of the new methods is baked into the design, even though it actually contradicts the docs (which never mention double refresh...) Given that it took 4 months to get revalidatePath to work on dynamic routes, see https://github.com/vercel/next.js/issues/49387, and the great number of other issues with ISR in app directory starting with 13.4, it seems pretty unclear what and when will happen next. Sorry I can't assist with actual coding.

**Reply by medonat** [reply]
Thank you for pointing out that the docs are not correct. I lost hours as I thought our implementation of On demand revalidation is broken. I am almost sure that it worked with a single refresh in 13.4.7.

**Reply by medonat** [reply]
We updated to latest version (13.5.5) and it seems fixed.

**Reply by jwalcher** [reply]
Agreed, the double refresh is not necessary any longer. I'm still experiencing issues with  revalidatePath on routes that are rewritten by middleware, but that's for another time.
